### PR TITLE
Assignment 2

### PR DESCRIPTION
**Pull request by katmilton**
I worked with @thigiang16.

### CODE REVIEW

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -26,14 +22,25 @@ public void Error_WithData_LogsMessage()
...
         // Assert
         Assert.AreEqual(1, logger.LoggedMessages.Count);
         Assert.AreEqual(LogLevel.Error, logger.LoggedMessages[0].LogLevel);
         Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
     }
 
+    [TestMethod]
+    public void Warning_UsesCorrectLogLevelandFormats_Success()
+    {
+        var logger = new TestLogger();
```
Comment (KorbinWeiler) [inline Logger.Tests/BaseLoggerMixinsTests.cs]: var should be replaced with the correct type to reduce ambiguity and increase code readability.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 2]
```diff
@@ -26,14 +22,25 @@ public void Error_WithData_LogsMessage()
...
+        logger.Error("Message {0}", 42);
 
         // Assert
         Assert.AreEqual(1, logger.LoggedMessages.Count);
         Assert.AreEqual(LogLevel.Error, logger.LoggedMessages[0].LogLevel);
         Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
     }
 
+    [TestMethod]
+    public void Warning_UsesCorrectLogLevelandFormats_Success()
```
Comment (KorbinWeiler) [inline Logger.Tests/BaseLoggerMixinsTests.cs]: It would probably be better to split this into two separate tests to check the correct log level and format separately. This will reduce confusion if only one of the two parts fails.

Comment (KorbinWeiler) [inline Logger.Tests/BaseLoggerMixinsTests.cs]: You are only testing that the Warning works. You should also have tests to make sure that Error, Debug, and Information work as well.

File: Logger/Logger.csproj [hunk 3]
```diff
@@ -1,8 +1,13 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>net9.0</TargetFramework>
     <LangVersion>11.0</LangVersion>
```
Comment (KorbinWeiler) [inline Logger/Logger.csproj]: The most recent language version is 13.

**Review by KorbinWeiler — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to store class name - the name of the class that calls the LogFactory's CreateLogger() method. This class (that calls the CreateLogger() method) will identify its name using nameof(), as specified below.) This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.
Extra Credit
Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌
Implement the factory pattern using static abstract methods on a Logger interface instead.❌

File: Logger/Logger.csproj (hunk 3 above)
Comment (thigiang16) [reply Logger/Logger.csproj]: Thank you for pointing it out. I just fixed it

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 2 above)
Comment (thigiang16) [reply Logger.Tests/BaseLoggerMixinsTests.cs]: Thank you for the suggestion. I split it into two separate tests

**Review by Hellothereyoko — APPROVED** [review]
How were you able to do that without expressed declaration?? I tried for like 3 hrs, you gotta teach me!

File: Logger/FileLogger.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.IO;
+
+namespace Logger
+{
+    public class FileLogger : BaseLogger
+    {
+        private readonly string _filePath;
```
Comment (PigPig-SDK) [inline Logger/FileLogger.cs]: I believe PascalCase should be used.
_filePath -> _FilePath

File: Logger/BaseLoggerMixins.cs [hunk 5]
```diff
@@ -1,6 +1,30 @@
...
+        => LogWithLevel(logger, LogLevel.Information, message, args);
+
+    public static void Debug(this BaseLogger logger, string message, params object[] args)
+        => LogWithLevel(logger, LogLevel.Debug, message, args);
+
+    private static void LogWithLevel(BaseLogger logger, LogLevel level, string message, object[] args)
+    {
+       ArgumentNullException.ThrowIfNull(logger);
+
+        var formatted = (args is { Length: > 0 })
```
Comment (chanderlud) [inline Logger/BaseLoggerMixins.cs]: For this logic, I think it is fine to pass an empty args array into the `string.Format` method, so there is no need to check if args is empty. Let me know if I'm missing something here. But I just did this simpler logic:
```
logger.Log(level, string.Format(CultureInfo.InvariantCulture, message, args));
```

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 6]
```diff
@@ -26,22 +22,115 @@ public void Error_WithData_LogsMessage()
...
+        TestLogger logger = new TestLogger();
+
+        logger.Warning("Low disk: {0}%", 12);
+
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
+        Assert.AreEqual("Low disk: 12%", logger.LoggedMessages[0].Message);
     }
+
+    [TestMethod]
+    public void Error_UsesCorrectLogLevel_Succes()
```
Comment (chanderlud) [inline Logger.Tests/BaseLoggerMixinsTests.cs]: Just a minor typo here, should be `Error_UsesCorrectLogLevel_Success`

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 7]
```diff
@@ -26,22 +22,115 @@ public void Error_WithData_LogsMessage()
...
-    public override void Log(LogLevel logLevel, string message)
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
+        Assert.AreEqual(LogLevel.Warning, logger.LoggedMessages[0].LogLevel);
+    }
+
+    [TestMethod]
+    public void Warning_FormatsMessageCorrectly_Success()
     {
-        LoggedMessages.Add((logLevel, message));
+        TestLogger logger = new TestLogger();
```
Comment (chanderlud) [inline Logger.Tests/BaseLoggerMixinsTests.cs]: There are a number of places in your tests where the new expression can be simplified. For example, `TestLogger logger = new();` in this case.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 8]
```diff
@@ -26,22 +22,115 @@ public void Error_WithData_LogsMessage()
...
+    {
+        TestLogger logger = new TestLogger();
 
-public class TestLogger : BaseLogger
-{
-    public List<(LogLevel LogLevel, string Message)> LoggedMessages { get; } = new List<(LogLevel, string)>();
+        logger.Warning("Low disk: {0}%", 12);
 
-    public override void Log(LogLevel logLevel, string message)
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
```
Comment (chanderlud) [inline Logger.Tests/BaseLoggerMixinsTests.cs:40]: The assertion `Assert.HasCount` is suggested as a more appropriate choice when checking the length of an array. For example, it would be `Assert.HasCount(1, logger.LoggedMessages);` here. There are a number of instances of this throughout your tests.

File: Logger/BaseLoggerMixins.cs [hunk 9]
```diff
@@ -1,6 +1,30 @@
-﻿namespace Logger;
+﻿using System;
+using System.Globalization;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger logger, string message, params object[] args)
```
Comment (chanderlud) [inline Logger/BaseLoggerMixins.cs:8]: I think that all of these Mixin methods should have a signature to accept a nullable BaseLogger; this will also fix your test `Error_WithNullLogger_ThrowsException` where you currently need to use `null!` to force null into the method.

**Review by PigPig-SDK — APPROVED** [review]
There is an existing BaseLogger class. It needs an auto property to store class name - the name of the class that calls the LogFactory's CreateLogger() method. This class (that calls the CreateLogger() method) will identify its name using nameof(), as specified below.) This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

**Review by chanderlud — COMMENTED** [review]
Assignment 2 Peer Review checklist:

- There is an existing BaseLogger class. It needs an auto property to store class name - the name of the class that calls the LogFactory's CreateLogger() method. This class (that calls the CreateLogger() method) will identify its name using nameof(), as specified below.) This property should be set in the LogFactory using an object initializer. ✔

- Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"

- The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔

- If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔

Inside of BaseLoggerMixins implement extension methods on BaseLogger for

  Error, ✔
  Warning, ✔
  Information, and ✔
  Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.

- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
- Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

Looks like everything is working & tested well. I left a few comments for minor changes that I think will improve your code quality.

**Comment by JosephPotapenko** [comment]
Assignment
There is an existing BaseLogger class. It needs an auto property to store class name - the name of the class that calls the LogFactory's CreateLogger() method. This class (that calls the CreateLogger() method) will identify its name using nameof(), as specified below.) This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

**Comment by JosephPotapenko** [comment]
The LogWithLevel method checks if args has elements before calling string.Format().  
But string.Format already deals with argument arrays, so the check isn't really necessary.

```csharp
private static void LogWithLevel(BaseLogger? logger, LogLevel level, string message, params object[] args)
{
    ArgumentNullException.ThrowIfNull(logger);
    var formatted = string.Format(CultureInfo.InvariantCulture, message, args);
    logger.Log(level, formatted);
}
```

In BaseLoggerMixinTests.cs,  public void Error_UsesCorrectLogLevel_Succes() is missing an s in line 57.

```csharp
 public void Error_UsesCorrectLogLevel_Success()
```

You announce the tests and have a slightly redundant type declaration in several tests. TestLogger logger = new TestLogger();
This can be simplified.

```csharp
TestLogger logger = new();
```

Otherwise, I feel like most of the other issues were covered previously. Looking good if you fix that all.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 2 above)
Comment (thigiang16) [reply Logger.Tests/BaseLoggerMixinsTests.cs]: Thanks for the feedback! I added more tests for Error, Debug, and Information.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 7 above)
Comment (thigiang16) [reply Logger.Tests/BaseLoggerMixinsTests.cs]: Thanks for pointing that out! I fixed it

File: Logger/BaseLoggerMixins.cs (hunk 5 above)
Comment (thigiang16) [reply Logger/BaseLoggerMixins.cs]: Your version is definitely simpler. I'll update mine

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 8 above)
Comment (thigiang16) [reply Logger.Tests/BaseLoggerMixinsTests.cs:40]: I see what you mean, thanks for pointing that out! I'll go through my tests and check for them

File: Logger/BaseLoggerMixins.cs (hunk 9 above)
Comment (thigiang16) [reply Logger/BaseLoggerMixins.cs:8]: Thanks for catching that!

Comment (Joshua-Lester3) [reply Logger/BaseLoggerMixins.cs:8]: @chanderlud Going along with our lecture yesterday, just wanted to note that making the `BaseLogger` param nullable would be bad here. Our intent is that it should not be null, since we are adding a check that it is not null. Having it be nullable would be saying "it's okay for this param to be null" and then punishing the user for passing null by throwing an exception. Not a big deal as this was before our lecture.

A funny analogy Ben gave in my class: it's like saying "come over here, I have some candy for you" and then kicking them in the teeth when they come over. Probably getting that wrong, but similar energy. It's just saying that telling someone that this *thing* is okay, then punishing them for doing that *thing*.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 10]
```diff
@@ -8,15 +8,11 @@ namespace Logger.Tests;
 public class BaseLoggerMixinsTests
 {
     [TestMethod]
-    [ExpectedException(typeof(ArgumentNullException))]
+    //[ExpectedException(typeof(ArgumentNullException))]
```
Comment (Joshua-Lester3) [inline Logger.Tests/BaseLoggerMixinsTests.cs:11]: I'd remove all unused code/comments.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 8 above)
Comment (Joshua-Lester3) [reply Logger.Tests/BaseLoggerMixinsTests.cs:40]: @thigiang16 just a reminder to follow through on comments - totally get maybe something came up after you commented that and you didn't get to it. So maybe implement the changes, if you agree with them, and then comment that they're done. Or if you don't agree with the review comment or want to continue the discussion, reply back with your thoughts.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 11]
```diff
@@ -26,22 +22,115 @@ public void Error_WithData_LogsMessage()
...
+    [TestMethod]
+    public void Error_FormatsMessageCorrectly_Success()
+    { 
+        TestLogger logger = new();
+
+        logger.Error("System failure: {0}", "Disk error");
+
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
+        Assert.AreEqual("System failure: Disk error", logger.LoggedMessages[0].Message);
+    }
```
Comment (Joshua-Lester3) [inline Logger.Tests/BaseLoggerMixinsTests.cs:76]: @BenjaminMichaelis I'm curious your thoughts on this. Would it be better to combine these into one test? I remember someone saying that you should only be testing one thing in a test. But I'd argue that these two things are a part of the same behavior, and could be combined.

File: Logger.Tests/FileLoggerTests.cs [hunk 12]
```diff
@@ -1,9 +1,53 @@
...
+using System.IO;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Logger.Tests;
 
 [TestClass]
 public class FileLoggerTests
 {
+    private readonly string _testFilePath = "testlog.txt";
```
Comment (Joshua-Lester3) [inline Logger.Tests/FileLoggerTests.cs:12]: A [guideline](https://essentialcsharp.com/guidelines#properties) I'd say still applies in testing:
> DO favor automatically implemented properties over fields.

File: Logger.Tests/LogFactoryTests.cs [hunk 13]
```diff
@@ -6,4 +6,26 @@ namespace Logger.Tests;
 public class LogFactoryTests
 {
 
+    [TestMethod]
+    public void CreateLogger_WithoutConfiguration_ReturnsNull()
+    {
+        var factory = new Logger.LogFactory();
+        var logger = factory.CreateLogger(nameof(LogFactoryTests));
```
Comment (Joshua-Lester3) [inline Logger.Tests/LogFactoryTests.cs:13]: [Guideline](https://essentialcsharp.com/guidelines#types):
> AVOID using implicitly typed local variables ( var ) unless the data type of the assigned value is obvious.

File: Logger/BaseLogger.cs [hunk 14]
```diff
@@ -3,5 +3,7 @@
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string ClassName { get; init; } = string.Empty;
```
Comment (Joshua-Lester3) [inline Logger/BaseLogger.cs:7]: I'd probably implement a check here in the property to ensure it's not null. Similar to the example Ben showed yesterday using a nullable backing field, with null checks upon setting (or initting), and setting in the constructor.

File: Logger/FileLogger.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.IO;
+
+namespace Logger
+{
+    public class FileLogger : BaseLogger
+    {
+        private readonly string _FilePath;
```
Comment (Joshua-Lester3) [inline Logger/FileLogger.cs:8]: I'd favor properties over fields, like I noted before. [Guideline](https://essentialcsharp.com/guidelines#properties):
> DO favor automatically implemented properties over fields.

File: Logger/FileLogger.cs [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+        public FileLogger(string filePath) 
+        {
+            _FilePath = filePath;
```
Comment (Joshua-Lester3) [inline Logger/FileLogger.cs:12]: I'd do a null check/empty check, either here or in the property setter/initer (if that was changed above).

File: Logger/LogFactory.cs [hunk 17]
```diff
@@ -2,9 +2,24 @@
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _filePath;
+
+    public void ConfigureFileLogger(string filePath)
+    {
+        _filePath = filePath;
```
Comment (Joshua-Lester3) [inline Logger/LogFactory.cs:9]: I'd also do a null/empty check here.

File: Logger/LogFactory.cs [hunk 18]
```diff
@@ -2,9 +2,24 @@
...
+    }
+
+    public BaseLogger? CreateLogger(string className)
     {
 
-        return null;
+        if (string.IsNullOrWhiteSpace(_filePath))
+        {
```
Comment (Joshua-Lester3) [inline Logger/LogFactory.cs:16]: And a null/empty check here for `className`.

File: Logger/LogFactory.cs [hunk 19]
```diff
@@ -2,9 +2,24 @@
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _filePath;
```
Comment (Joshua-Lester3) [inline Logger/LogFactory.cs:5]: Favor properties here, also [naming convention](https://essentialcsharp.com/guidelines#naming-conventions-table) should be `_FilePath`.

File: Logger.Tests/Logger.Tests.csproj [hunk 20]
```diff
@@ -1,9 +1,14 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
-		<TargetFramework>net7.0</TargetFramework>
+		<TargetFramework>net9.0</TargetFramework>
 		<IsPackable>false</IsPackable>
 		<LangVersion>11.0</LangVersion>
```
Comment (Joshua-Lester3) [inline Logger.Tests/Logger.Tests.csproj:6]: Needs to be latest version

**Review by Joshua-Lester3 — COMMENTED** [review]
## Assignment

- There is an existing `BaseLogger` class. It needs an **auto property** to store class name - the name of the class that calls the `LogFactory`'s `CreateLogger()` method. This class (that calls the `CreateLogger()` method) will identify its name using nameof(), as specified below. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
    Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ✔
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ❌ LangVersion not latest release in Logger.Tests: 13.0
- Turn on Nullability (`Nullable`) ✔
- **All of the above should be unit tested.** ✔

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ❌ Not implemented
- Implement the factory pattern using static abstract methods on a Logger interface instead. ❌ Not implemented
